VideoMediaSampleRenderer will produce out of order, corrupted frame when used with b-frames.

jyavenard opened the pull request:
#### ad4c63c079db6dd47dd2487858aae1a850e64bbf
<pre>
VideoMediaSampleRenderer will produce out of order, corrupted frame when used with b-frames.
<a href="https://bugs.webkit.org/show_bug.cgi?id=283235">https://bugs.webkit.org/show_bug.cgi?id=283235</a>
<a href="https://rdar.apple.com/140046888">rdar://140046888</a>

Reviewed by Youenn Fablet.

1- The queue used to feed the decoder was a CMBufferQueue which re-order samples in presentation order.
So we ended up feeding the decoder the frames in presentation order rather than decode order, leading to decoding artifacts
2- The way the WebCoreDecompression was used by the VideoMediaSampleRenderer, was to immediately queue the decoded frame to the AVSampleBufferDisplayLayer,
without letting the CMBufferQueue re-sort the frames by presentation order. So frames were displayed in decode order instead.

To solve 1-, we remove the use of a sorted CMBufferQueue and instead use a plain Dequeue. The Dequeue store the compressed
image along the flushing session id. Whenever we process the compressed frame, if the Ids don&apos;t match, the frame is dropped.
For 2- we only add the decoded frame at the time it is ready for display, using the AV synchroniser&apos;s CMTimeBase timer.
By only ever dequeuing the video frame at the time we should be displaying it, we guarantee that the frames have been fully
sorted in presentation order, without the downside of added latency (as we would otherwise have to wait until we&apos;ve reached the
minimum required number of frames in the queue: the &quot;H264 sliding window&quot;).
If we were to display a frame out of order, those future incoming earlier frames would have been dropped anyway by the time they are received
being now too late.

Fly-by: setTimeBase if called multiple times, could potentially cause access to the m_timebase member to be access concurrently.
We limit the use so that it can be either set once or cleared and protect the variable member with a lock.

Manually tested. The VideoMediaSampleRenderer isn&apos;t currently used with players supporting B-Frames.

* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.h:
(WebCore::VideoMediaSampleRenderer::timebase const): Deleted.
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
(WebCore::VideoMediaSampleRenderer::~VideoMediaSampleRenderer): Cancel (flush) any pending decoding tasks.
(WebCore::VideoMediaSampleRenderer::isReadyForMoreMediaData const):
(WebCore::VideoMediaSampleRenderer::maybeBecomeReadyForMoreMediaData):
(WebCore::VideoMediaSampleRenderer::setTimebase): Whenever the CMTimeSource calls, maybe enqueue for display the earliest decoded frame.
(WebCore::VideoMediaSampleRenderer::clearTimebase):
(WebCore::VideoMediaSampleRenderer::timebase const):
(WebCore::VideoMediaSampleRenderer::enqueueSample):
(WebCore::VideoMediaSampleRenderer::decodeNextSample):
(WebCore::VideoMediaSampleRenderer::initializeDecompressionSession): Fly-by: the current code can&apos;t deal with m_decodedSampleQueue changing mid-flight
as accessed to this member would become racy. As such, we make it an error to clear m_decodedSampleQueue.
There&apos;s also no need to set the timebase and ResourceOwner to the WebCoreDecompression as we aren&apos;t using its internal samples queueing.
(WebCore::VideoMediaSampleRenderer::decodedFrameAvailable):
(WebCore::VideoMediaSampleRenderer::maybeQueueFrameForDisplay):
(WebCore::VideoMediaSampleRenderer::flushCompressedSampleQueue):
(WebCore::VideoMediaSampleRenderer::flushDecodedSampleQueue):
(WebCore::VideoMediaSampleRenderer::cancelTimer):
(WebCore::VideoMediaSampleRenderer::purgeDecodedSampleQueue):
(WebCore::VideoMediaSampleRenderer::flush):
(WebCore::VideoMediaSampleRenderer::expectMinimumUpcomingSampleBufferPresentationTime): Fly-by: When using a decompression session
frames queued for display are always earlier than the value set in expectMinimumUpcomingSampleBufferPresentationTime as
the time given if of the compressed frame about to be enqueued leading to very AVFoundation verbose warnings.
(WebCore::VideoMediaSampleRenderer::copyDisplayedPixelBuffer): Fly-By, if we don&apos;t have an active WebCoreDecompressionSession and [AVSBDL copyDisplayedPixelBuffer]
didn&apos;t return an image, exit early as there&apos;s nothing to retrieve in the decoded queue.
(WebCore::VideoMediaSampleRenderer::copyDisplayedPixelBuffer): add workQueue assertion.

Canonical link: <a href="https://commits.webkit.org/286719@main">https://commits.webkit.org/286719@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8adaf8988f9c4f3201ea0d3ce37bd83ec46b4358

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/76925 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/55960 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/29831 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/81469 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/28201 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/79042 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/65102 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4253 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/60278 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18361 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/79992 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50228 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/66037 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40593 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47630 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23532 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/26527 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68748 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23863 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/82905 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/4301 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2873 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68553 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/4455 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66010 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67811 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16893 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11798 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9879 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4248 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/7061 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/4271 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/7702 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/6029 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.h:
```diff
@@ -95,24 +96,32 @@ class VideoMediaSampleRenderer : public ThreadSafeRefCountedAndCanMakeThreadSafe
 
     void resetReadyForMoreSample();
     void initializeDecompressionSession();
+    void enqueueSample(RetainPtr<CMSampleBufferRef>&&, int);
```
I don't see this one as implemented.

jyavenard replied:
removed

youennf commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -178,35 +202,50 @@ - (void)resetUpcomingSampleBufferPresentationTimeExpectations;
+        if (RefPtr protectedThis = weakThis.get()) {
+            assertIsCurrent(protectedThis->m_workQueue.get());
+
+            if (flushId != protectedThis->m_flushId) {
```
I am not sure to understand why we need `m_flushId`.
It seems `flush()` could simply clear `m_compressedSampleQueue` by hopping to the work queue.
Why do we need to be that aggressive?

Similarly, on the rendering side, I would expect flushing the decompression session to properly handle flush without needed a m_flushId here.

Do we even need to use a decompression session here?
It seems we should either fix decompression session to do more for VideoMediaSampleRenderer or have VideoMediaSampleRenderer use a VideoDecoder directly (once we migrate VT decoder support to WebCore/VideoDecoder).

jyavenard replied:
we don't want any of the frames to be decoded, nor those in the process of being decoded to be added to the queue. It also prevents blocking the main thread while we wait for a decode to complete.

Ultimately, yes, all that does similar in WebCoreDecompressionSession will be removed and we will only use `VideoMediaSampleRenderer` 

Having this in mind, it was easier to have the `VideoMediaSampleRenderer` manage the reordering of the decoding samples than have the WebCoreDecompressionSession do it. The WebCoreDecompressionSession doesn't have a timebase set , it decodes one sample at a time. So for the `WebCoreDecompressionSession` to do the re-ordering, it would need to keep the frames in the queue until the sliding window has been fully received adding latency.

As the VideoMediaSampleRenderer can simply pull the frame from the re-ordering queue only when it's actually time to use the frame, it guarantees that the frames are in the right order without latency.

I hinted at this in the commit log

jyavenard replied:
oh, m_flushId is increased whenever `flush()` is called

youennf commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -178,35 +202,50 @@ - (void)resetUpcomingSampleBufferPresentationTimeExpectations;
 
+    auto [sample, flushId] = m_compressedSampleQueue.takeFirst();
+    if (flushId != m_flushId)
+        return decodeNextSample();
```
I am not clear why we need this check if we clear  m_compressedSampleQueue as part of `flush()`

jyavenard replied:
we dispatch a task to clear m_compressedSampleQueue; but an earlier call to `enqueueSample()` could be on the way.  We could also already have a few compressed sample in the queue. 

This allows to not necessarily decode anything between the time we call flush() and the time `m_compressedSampleQueue.clear()` is called.

youennf commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -178,35 +202,50 @@ - (void)resetUpcomingSampleBufferPresentationTimeExpectations;
 
         protectedThis->m_isDecodingSample = false;
+        if (flushId != protectedThis->m_flushId)
+            return;
```
I would expect m_decompressionSession flush to handle this somehow.
If m_decompressionFlush is called, then the promise is resolved with no result and we try decoding the next sample.
Would that be simpler?

jyavenard replied:
It would be, but I hinted at this in the commit log. Making WebCoreDecompressionSession return in PTS order frame, we need to increase its latency, and we wouldn't be able to keep the simple 1:1 ratio between call to decode and having a one frame out

youennf commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -281,41 +342,49 @@ - (void)resetUpcomingSampleBufferPresentationTimeExpectations;
-
-    PAL::CMBufferQueueReset(m_compressedSampleQueue.get());
-    PAL::CMTimebaseSetTimerDispatchSourceNextFireTime(m_timebase.get(), m_timerSource.get(), PAL::kCMTimeInvalid, 0);
+    ++m_flushId;
```
Ah I see, m_flushId is called at the same time decompression session flush is called.
The scheme I was mentioning above would require flushing the decompressions session from the work queue.

There might still be room for simplification but I am not exactly sure how then.

jyavenard replied:
yes, the flushId is a scheme I've used quite a bit in similar case. It's simple, avoid the need for sync dispatch nor the use of a lock.

And it makes it much easier to fully dismiss any in-flight operation where we can dismiss the result.

We could have achieved similar cancellation of inflight decoding by using a NativePromise::Request , but this ended up the need of a lock which I wanted to avoid.

youennf commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.h:
```diff
@@ -95,24 +96,32 @@ class VideoMediaSampleRenderer : public ThreadSafeRefCountedAndCanMakeThreadSafe
+    RetainPtr<CMBufferQueueRef> m_decodedSampleQueue WTF_GUARDED_BY_CAPABILITY(m_workQueue.get());
     RefPtr<WebCoreDecompressionSession> m_decompressionSession;
     bool m_isDecodingSample { false };
+    std::atomic<bool> m_isDisplayingSample { false };
```
Isn't this one always checked/set from the work queue like m_decodedSampleQueue?

jyavenard replied:
yes indeed. it's a remnant of an earlier code where it was checked before the call to copyPixelBuffer to avoid the sync dispatch. I'll remove it. That code is re-introduced in another PR

youennf commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -132,23 +135,44 @@ - (void)resetUpcomingSampleBufferPresentationTimeExpectations;
 
 void VideoMediaSampleRenderer::setTimebase(RetainPtr<CMTimebaseRef>&& timebase)
```
Maybe we should add clearTimebase() and expect setTimebase to only be called when m_timebase is null.

jyavenard replied:
yes, good point that may make it more clear, it could be private too so it's even more clear that you can only set the timebase once and avoid making usage error

youennf commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -178,35 +202,50 @@ - (void)resetUpcomingSampleBufferPresentationTimeExpectations;
-        if (RefPtr protectedThis = weakThis.get())
+    ++m_framesBeingDecoded;
+    m_workQueue->dispatch([weakThis = ThreadSafeWeakPtr { *this }, sample = RetainPtr { cmSampleBuffer }, flushId = m_flushId.load()]() mutable {
+        if (RefPtr protectedThis = weakThis.get()) {
```
We tend to prefer `if (!protectedThis) return`

youennf commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.h:
```diff
@@ -95,24 +96,32 @@ class VideoMediaSampleRenderer : public ThreadSafeRefCountedAndCanMakeThreadSafe
+    Deque<std::pair<RetainPtr<CMSampleBufferRef>, int>> m_compressedSampleQueue WTF_GUARDED_BY_CAPABILITY(m_workQueue.get());
+    RetainPtr<CMBufferQueueRef> m_decodedSampleQueue WTF_GUARDED_BY_CAPABILITY(m_workQueue.get());
     RefPtr<WebCoreDecompressionSession> m_decompressionSession;
     bool m_isDecodingSample { false };
```
Should it be guarded by capability with m_workQueue as well?

youennf commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -132,23 +135,44 @@ - (void)resetUpcomingSampleBufferPresentationTimeExpectations;
 
 void VideoMediaSampleRenderer::setTimebase(RetainPtr<CMTimebaseRef>&& timebase)
 {
+    Locker locker { m_lock };
```
If we no longer use m_timebase for WebCoreDecompressionSession, maybe we could only touch m_timebase in m_workQueue and not use the lock for m_timebase?

jyavenard replied:
> If we no longer use m_timebase for WebCoreDecompressionSession, maybe we could only touch m_timebase in m_workQueue and not use the lock for m_timebase?

we do access `m_timerSource` from the main thread to cancel the timer in flush. I can't see a way to do so without a lock as the callback is an objC-block.

jyavenard replied:
in `clearTimer()` method

youennf reviewed (approved):
LGTM, but I wonder whether we could not simplify a bit things.
Also, is the long term goal to stop using decompression session for VideoMediaSampleRenderer (and instead use a VideoDecoder?)

jyavenard committed:
VideoMediaSampleRenderer will produce out of order, corrupted frame when used with b-frames.
https://bugs.webkit.org/show_bug.cgi?id=283235
rdar://140046888

Reviewed by Youenn Fablet.

1- The queue used to feed the decoder was a CMBufferQueue which re-order samples in presentation order.
So we ended up feeding the decoder the frames in presentation order rather than decode order, leading to decoding artifacts
2- The way the WebCoreDecompression was used by the VideoMediaSampleRenderer, was to immediately queue the decoded frame to the AVSampleBufferDisplayLayer,
without letting the CMBufferQueue re-sort the frames by presentation order. So frames were displayed in decode order instead.

To solve 1-, we remove the use of a sorted CMBufferQueue and instead use a plain Dequeue. The Dequeue store the compressed
image along the flushing session id. Whenever we process the compressed frame, if the Ids don't match, the frame is dropped.
For 2- we only add the decoded frame at the time it is ready for display, using the AV synchroniser's CMTimeBase timer.
By only ever dequeuing the video frame at the time we should be displaying it, we guarantee that the frames have been fully
sorted in presentation order, without the downside of added latency (as we would otherwise have to wait until we've reached the
minimum required number of frames in the queue: the "H264 sliding window").
If we were to display a frame out of order, those future incoming earlier frames would have been dropped anyway by the time they are received
being now too late.

Fly-by: setTimeBase if called multiple times, could potentially cause access to the m_timebase member to be access concurrently.
We limit the use so that it can be either set once or cleared and protect the variable member with a lock.

Manually tested. The VideoMediaSampleRenderer isn't currently used with players supporting B-Frames.

* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.h:
(WebCore::VideoMediaSampleRenderer::timebase const): Deleted.
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
(WebCore::VideoMediaSampleRenderer::~VideoMediaSampleRenderer): Cancel (flush) any pending decoding tasks.
(WebCore::VideoMediaSampleRenderer::isReadyForMoreMediaData const):
(WebCore::VideoMediaSampleRenderer::maybeBecomeReadyForMoreMediaData):
(WebCore::VideoMediaSampleRenderer::setTimebase): Whenever the CMTimeSource calls, maybe enqueue for display the earliest decoded frame.
(WebCore::VideoMediaSampleRenderer::clearTimebase):
(WebCore::VideoMediaSampleRenderer::timebase const):
(WebCore::VideoMediaSampleRenderer::enqueueSample):
(WebCore::VideoMediaSampleRenderer::decodeNextSample):
(WebCore::VideoMediaSampleRenderer::initializeDecompressionSession): Fly-by: the current code can't deal with m_decodedSampleQueue changing mid-flight
as accessed to this member would become racy. As such, we make it an error to clear m_decodedSampleQueue.
There's also no need to set the timebase and ResourceOwner to the WebCoreDecompression as we aren't using its internal samples queueing.
(WebCore::VideoMediaSampleRenderer::decodedFrameAvailable):
(WebCore::VideoMediaSampleRenderer::maybeQueueFrameForDisplay):
(WebCore::VideoMediaSampleRenderer::flushCompressedSampleQueue):
(WebCore::VideoMediaSampleRenderer::flushDecodedSampleQueue):
(WebCore::VideoMediaSampleRenderer::cancelTimer):
(WebCore::VideoMediaSampleRenderer::purgeDecodedSampleQueue):
(WebCore::VideoMediaSampleRenderer::flush):
(WebCore::VideoMediaSampleRenderer::expectMinimumUpcomingSampleBufferPresentationTime): Fly-by: When using a decompression session
frames queued for display are always earlier than the value set in expectMinimumUpcomingSampleBufferPresentationTime as
the time given if of the compressed frame about to be enqueued leading to very AVFoundation verbose warnings.
(WebCore::VideoMediaSampleRenderer::copyDisplayedPixelBuffer): Fly-By, if we don't have an active WebCoreDecompressionSession and [AVSBDL copyDisplayedPixelBuffer]
didn't return an image, exit early as there's nothing to retrieve in the decoded queue.
(WebCore::VideoMediaSampleRenderer::copyDisplayedPixelBuffer): add workQueue assertion.

Canonical link: https://commits.webkit.org/286719@main
